### ISSUE TITLE
Relationship between lws_callback_on_writable_all_protocol_vhost and LWS_CALLBACK_ESTABLISHED

### ISSUE THREAD

**Issue by ktb92677** [warmcat/libwebsockets#2197]
Hi, is there any relationship between lws_callback_on_writable_all_protocol_vhost and LWS_CALLBACK_ESTABLISHED? I am currently running into an issue where it is taking a really long time (around 25,000 MS) for a vhost (initialized during run time) to establish a new connection with a client. However oddly enough I noticed that if I run lws_callback_on_writable_all_protocol_vhost on an already existing vhost I initialized earlier in the program in a constant loop then for the same exact client code the LWS_CALLBACK_ESTABLISHED returns after around 2000 MS. I have tried a few things but these results are very consistent. I have tried naming the vhosts the same thing... different things, using the same and different callback functions. I don't know what else to think but that there must be some relationship between these two things.

**Comment by lws-team** [comment]
You need to reformulate your question into something that can be answered... something you don't really describe and don't show logs for takes 25s under some condition where you "created the vhost".

Describe the sequencing properly with some code if it will help.  Show detailed logs.  Check if same behaviour on latest lws or not.  Best of all is hack a minimal example to show the problem so I can reproduce it.

**Comment by ktb92677** [comment]
Here is a quick snippet of my libwebsockets wrapper that I use in my program:

```c
#include "WEB_SOCKET_SERVER.h"

lws_context	*WEB_SOCKET_SERVER::context = NULL;
thread		*WEB_SOCKET_SERVER::main_thread = NULL;
uint32_t	WEB_SOCKET_SERVER::refcount = 0;
uint8_t		WEB_SOCKET_SERVER::running = false;

int callback_http(
lws* wsi, lws_callback_reasons reason, void* user, void* in, size_t len) {
	(void) wsi;
	(void) reason;
	(void) user;
	(void) in;
	(void) len;
	return 0;
}

int callback_dumb_increment(lws* wsi, lws_callback_reasons reason,
	void* user, void* in, size_t len) {

	(void) user;

	WEB_SOCKET_ELEMENT *temp = NULL;
	uint32_t time_out = 0;
	WEB_SOCKET_SERVER *web_socket_server = (WEB_SOCKET_SERVER*)
		lws_vhost_user(lws_get_vhost(wsi));

	switch (reason) {
		case LWS_CALLBACK_ESTABLISHED:
			web_socket_server->accept = true;
			break;
		case LWS_CALLBACK_SERVER_WRITEABLE:
			if (web_socket_server->active_write->size() <= 0) {
				break;
			}

			web_socket_server->active_write->pop(temp);

			lws_write(wsi, (unsigned char*)
			&temp->bytes[LWS_SEND_BUFFER_PRE_PADDING], temp->size,
			LWS_WRITE_BINARY);

			temp->size = -1;

			if (web_socket_server->active_write->size() > 0) {
				lws_callback_on_writable_all_protocol_vhost(
				lws_get_vhost(wsi), lws_get_protocol(wsi));
			}

			web_socket_server->pool->push(temp);
			break;
		case LWS_CALLBACK_RECEIVE:
			while (web_socket_server->pool->size() <= 0 &&
				time_out < WEB_SOCKET_TIME_OUT) {

				time_out++;
				Sleep_Milli(WEB_SOCKET_SLEEP_TIME);
			}
			if (time_out >= WEB_SOCKET_TIME_OUT) {
				break;
			}

			web_socket_server->pool->pop(temp);

			if (len > MAX_WEBSOCKET_PACKET_SIZE) {
				log_err("received large packet. truncated.");
				len = MAX_WEBSOCKET_PACKET_SIZE;
			}

			copy((uint8_t*)in, ((uint8_t*)in) + len, temp->bytes);
			temp->size = len;

			web_socket_server->active_read->push(temp);
            		break;
		case LWS_CALLBACK_CLOSED:
			web_socket_server->disconnected = true;
			break;
    }
    return 0;
}

bool Initialize_WEB_SOCKET_SERVER(WEB_SOCKET_SERVER *web_socket_server) {
	if (web_socket_server->refcount <= 0) {
		log_dbg("initializing master websocket context");

		lws_context_creation_info context_info;
		memset(&context_info, 0, sizeof(context_info));

		context_info.options = LWS_SERVER_OPTION_EXPLICIT_VHOSTS;

		web_socket_server->context = lws_create_context(&context_info);
		if (web_socket_server->context == NULL) {
			log_err("failed to init master websocket context");
        		return false;
		}

		web_socket_server->running = true;
		web_socket_server->vhost = NULL;
		web_socket_server->accept = false;
		web_socket_server->main_thread = new thread(
			Service_Thread_WEB_SOCKET_SERVER, web_socket_server);
	}
	web_socket_server->refcount++;
	log_dbg("initializing new web_socket_server");

	WEB_SOCKET_ELEMENT *temp;

	web_socket_server->pool		= new Queue<WEB_SOCKET_ELEMENT*>;
	web_socket_server->active_write	= new Queue<WEB_SOCKET_ELEMENT*>;
	web_socket_server->active_read	= new Queue<WEB_SOCKET_ELEMENT*>;

	for (int i = 0; i < WEB_SOCKET_POOL_SIZE; i++) {
		temp		= new WEB_SOCKET_ELEMENT;
		temp->size	= -1;
		temp->bytes	= new uint8_t[MAX_WEBSOCKET_PACKET_SIZE];

		web_socket_server->pool->push(temp);
	}

	return true;
}

void Service_Thread_WEB_SOCKET_SERVER(WEB_SOCKET_SERVER *web_socket_server) {
	while (web_socket_server->running) {
		lws_service(web_socket_server->context, WEB_SOCKET_SLEEP_TIME);

		if (web_socket_server->vhost != NULL &&
			web_socket_server->accept) {

			lws_callback_on_writable_all_protocol_vhost(
				web_socket_server->vhost,
				lws_vhost_name_to_protocol(
				web_socket_server->vhost,
				"dumb-increment-protocol"));
		}
	}
}

bool Listen_WEB_SOCKET_SERVER(WEB_SOCKET_SERVER *web_socket_server,
	uint32_t port) {

	lws_context_creation_info context_info;
	memset(&context_info, 0, sizeof(context_info));

	web_socket_server->accept	= false;
	web_socket_server->disconnected	= false;
	DEFAULT_PROTOCOLS(web_socket_server->protocols);

	context_info.port = port;
	context_info.protocols = web_socket_server->protocols;
	context_info.gid = -1;
	context_info.uid = -1;
	context_info.user = (void*) web_socket_server;

	web_socket_server->vhost = lws_create_vhost(web_socket_server->context,
		&context_info);

	if (web_socket_server->vhost == NULL) {
		log_err("failed to initialize web socket server");
		return false;
	}

	while (!web_socket_server->accept) {
		Sleep_Milli(WEB_SOCKET_SLEEP_TIME);
	}

	return true;
}
```

The snippet of code that is concerning is as follows:

```c
lws_callback_on_writable_all_protocol_vhost(
web_socket_server->vhost,
lws_vhost_name_to_protocol(
web_socket_server->vhost,
"dumb-increment-protocol"));
```

Without this statement inside of the loop I notice that any vhost I initialize mid-way through the program take forever to get to LWS_CALLBACK_ESTABLISHED. I hope that is enough context?

**Comment by lws-team** [comment]
Please directly reference the copious amount of minimal examples instead of this kind of thing, they show how to get stuff working well and are licensed so you can cut and paste them even in your proprietary code

```
void Service_Thread_WEB_SOCKET_SERVER(WEB_SOCKET_SERVER *web_socket_server) {
	while (web_socket_server->running) {
		lws_service(web_socket_server->context, WEB_SOCKET_SLEEP_TIME);

		if (web_socket_server->vhost != NULL &&
			web_socket_server->accept) {

			lws_callback_on_writable_all_protocol_vhost(
				web_socket_server->vhost,
				lws_vhost_name_to_protocol(
				web_socket_server->vhost,
				"dumb-increment-protocol"));
		}
	}
}
```

```
	while (!web_socket_server->accept) {
		Sleep_Milli(WEB_SOCKET_SLEEP_TIME);
	}
```

depending on your platform details, one thread may not see changes another thread made until you do a memory barrier operation, or that part of the cache is flushed by something else.

If you want a ws server, literally start with this:

https://libwebsockets.org/git/libwebsockets/tree/minimal-examples/ws-server/minimal-ws-server

and add the vhost and see if you have a 25s delay magically coming.  Don't add your threads and magic sleep loops.

**Comment by ktb92677** [comment]
So I have added extensive print statements throughout and this specific call back: "LWS_CALLBACK_ESTABLISHED" is being delayed considerably in the absence of a constant call to lws_callback_on_writable_all_protocol_vhost. Removing this while loop that you referenced also produces the same result:

```c
while (!web_socket_server->accept) {
		Sleep_Milli(WEB_SOCKET_SLEEP_TIME);
	}
```

I figured you might know if there was some kind of odd synergy between "lws_callback_on_writable_all_protocol_vhost" and "LWS_CALLBACK_ESTABLISHED". Seeing as how that is not the case I will create a minimally reproducible example of this issue. I'll post back when I've created it. Thanks for all the help!
